scrapy/contrib/periodicstats/collectors.py

```diff
@@ -0,0 +1,157 @@
+            self.set_value(key=key, value=self.get_value(key), spider=spider)
+
+    def _setup_looping_task(self, _result=None, **kwargs):
+        spider = kwargs.pop('spider')
+        self._task = task.LoopingCall(self._process_interval_stats, spider)
+        d = self._task.start(1, **kwargs)
+        d.addErrback(self._setup_looping_task, spider=spider, now=False)
+
+    def _load_observers(self, crawler):
+        """
+        Loads the stats value observers
+        """
+        export_keys = set()
+        self._observers = defaultdict(list)
+        self._re_observers = defaultdict(list)
+        for observer in crawler.settings.get('PERIODIC_STATS_OBSERVERS', {}):
+            if not isinstance(observer, PeriodicStatsObserver):
+                raise NotConfigured('Stats observers must subclass PeriodicStatsObserver')
+            observer_export_key = observer.export_key
+            if observer_export_key in export_keys:
+                raise NotConfigured('Duplicate export key "%s"' % observer_export_key)
+            export_keys.add(observer_export_key)
+            if not observer.use_re_key:
+                self._observers[observer.key].append(observer)
+            else:
+                self._re_observers[observer.key].append(observer)
+
+    def _load_pipelines(self, crawler):
+        """
+        Loads the stats processor pipelines
+        """
+        self._pipelines = []
+        for pipeline in crawler.settings.get('PERIODIC_STATS_PIPELINES', []):
+            if not isinstance(pipeline, basestring):
+                raise NotConfigured('Stats pipelines must be defined as an object string path in settings')
+            pipeline_class = load_object(pipeline)
+            if not issubclass(pipeline_class, PeriodicStatsPipeline):
+                raise NotConfigured('Stats pipelines must subclass PeriodicStatsPipeline')
+            pipeline = pipeline_class.from_crawler(crawler)
+            self._pipelines.append(pipeline)
+
+    def _process_interval_stats(self, spider, is_close=False):
+        """
+        Returns the interval stats for the current interval
+        """
+        stats = self._get_interval_base_stats(spider=spider, is_close=is_close)
+        stats.update(self._get_interval_observers_stats(self._observers, force=is_close))
+        stats.update(self._get_interval_observers_stats(self._re_observers, force=is_close))
+        for pipeline in self._pipelines:
+            stats = pipeline.process_stats(spider=spider,
+                                           interval=self._interval+1,
+                                           stats=stats)
+            if not stats:
+                break
+        self._interval += 1
+        return stats
+
+    def _get_key_observers(self, key):
+        """
+        Returns a list of observers for the passed key
+        """
+        observers = []
+        for re_key, re_observers in self._re_observers.iteritems():
+            if re.match(re_key, key):
```

Comment:
I'd add the re.UNICODE flag just in case to support wild-carding unicode chars.
Ideally `key` could be an already compiled regex expression, but that kind of flexibility is probably not needed.